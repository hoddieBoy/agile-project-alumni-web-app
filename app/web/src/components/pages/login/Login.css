body {
    background-color: #f8f9fa;
    display: flex;
    flex-direction: column;
    min-height: 100vh;
    margin: 0;
}

header {
    flex-shrink: 0;
}

.main-content {
    flex: 1;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    padding: 10% 0;
}

footer {
    flex-shrink: 0;
    width: 100%;
    text-align: center;
    background-color: #002855;
    color: #ffffff;
    padding: 10px 0;
    position: absolute;
    bottom: 0;
}

.login-header {
    background-color: #002855;
    padding: 20px;
    color: #ffffff;
    text-align: center;
    position: relative;
}

.login-header img {
    position: absolute;
    top: 50%;
    left: 20px;
    transform: translateY(-50%);
    max-height: 120px;
}

.login-container {
    max-width: 35%;
    width: 100%;
    padding: 20px;
}

.btn-custom {
    background-color: #002855;
    color: #ffffff;
    border: none;
    padding: 10px 20px;
    border-radius: 4px;
    cursor: pointer;
    width: 100%;
    font-size: 16px;
}

.btn-custom:hover {
    background-color: #003366;
    color: #ffffff;
    border: 1px solid #002855;
}

.login-footer a {
    color: #ffffff;
    margin: 0 10px;
}

.form-group {
    margin-bottom: 15px;
}

.form-group label {
    font-weight: bold;
    color: #002855;
    display: block;
    margin-bottom: 5px;
}

.form-control {
    width: 100%;
    padding: 10px;
    border: 1px solid #ced4da;
    border-radius: 4px;
    box-sizing: border-box;
}

.form-control:focus {
    border-color: #002855;
    outline: none;
    box-shadow: 0 0 5px rgba(0, 40, 85, 0.3);
}
