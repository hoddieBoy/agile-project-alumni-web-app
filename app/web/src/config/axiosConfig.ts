import axios from 'axios';

const instance = axios.create(
    {
        baseURL: "http://localhost:8081/api/v1/alumni-fil",
        headers: {
            'Content-Type': 'application/json',
        }
    }
);

export default instance;