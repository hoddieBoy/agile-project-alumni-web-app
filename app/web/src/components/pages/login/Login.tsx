import React, {useState} from 'react';
import 'components/pages/login/Login.css';
import {Form, useActionData} from 'react-router-dom';

interface ErrorMessages {
    message: string;
}

function Login() {
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const error = useActionData() as ErrorMessages;

    return (
        <>
            <header className="login-header">
                <img
                    src="https://www.imt-atlantique.fr/sites/default/files/Images/Ecole/charte-graphique/IMT_Atlantique_logo_RVB_Negatif_Baseline_400x272.png"
                    alt="IMT Atlantique Logo"/>
                <h3>Welcome to Alumni FIL</h3>
                <p>Login to access administrative features</p>
            </header>

            <div className="main-content">
                <div className="login-container">
                    <Form method="post" id="login-form" action="/login">
                        <div className="form-group">
                            <label htmlFor="username">Username</label>
                            <input
                                type="text"
                                className="form-control"
                                id="username"
                                name="username"
                                placeholder="Enter your username"
                                value={username}
                                onChange={(e) => setUsername(e.target.value)}
                            />
                        </div>
                        <div className="form-group">
                            <label htmlFor="password">Password</label>
                            <input
                                type="password"
                                className="form-control"
                                id="password"
                                name="password"
                                placeholder="Enter your password"
                                value={password}
                                onChange={(e) => setPassword(e.target.value)}
                            />
                        </div>
                        {error?.message && <small className="danger">{error.message}</small>}
                        <button type="submit" className="btn btn-custom btn-block">Login</button>
                    </Form>
                </div>
            </div>

            <footer className="login-footer">
                <a href="#">Contact Us</a> | <a href="#">Privacy Policy</a>
            </footer>
        </>
    );
}

export default Login;
