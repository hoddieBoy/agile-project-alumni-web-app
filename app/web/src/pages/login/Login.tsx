import React, { useContext, useEffect, useState } from 'react';
import 'pages/login/Login.css';
import { Spin } from 'antd';
import { Form, Navigate, useActionData } from 'react-router-dom';
import Footer from "components/Footer";
import { AuthContext } from 'context/AuthContext';

interface LoginResponse {
    message: string;
    accessToken?: string;
    refreshToken?: string;
}

interface LoaderData {
    username: string;
}

function Login() {
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const [loading, setLoading] = useState(false);
    const response = useActionData() as LoginResponse;
    const { isAuthenticated,login } = useContext(AuthContext);

    useEffect(() => {
        setLoading(false);
        if (response?.accessToken && response?.refreshToken) {
            login(response.accessToken, response.refreshToken);
        }
    }, [response, login]);

    const handleSubmit = () => {
        setLoading(true);
    };

    if (isAuthenticated) {
        return <Navigate replace to="/search"/>;
    }

    return (
        <>
            <header className="login-header">
                <img
                    src="https://www.imt-atlantique.fr/sites/default/files/Images/Ecole/charte-graphique/IMT_Atlantique_logo_RVB_Negatif_Baseline_400x272.png"
                    alt="IMT Atlantique Logo"
                />
                <h3>Bienvenue sur Alumni FIL</h3>
                <p>Connectez-vous pour accéder aux fonctionnalités administratives</p>
            </header>

            <div className="main-content">
                <div className="login-container">
                    <Form method="post" id="login-form" action="/login" onSubmit={handleSubmit}>
                        <div className="form-group">
                            <label htmlFor="username">Nom d'utilisateur</label>
                            <input
                                type="text"
                                className="form-control"
                                id="username"
                                name="username"
                                placeholder="Entrez votre nom d'utilisateur"
                                value={username}
                                onChange={(e) => setUsername(e.target.value)}
                            />
                        </div>
                        <div className="form-group">
                            <label htmlFor="password">Mot de passe</label>
                            <input
                                type="password"
                                className="form-control"
                                id="password"
                                name="password"
                                placeholder="Entrez votre mot de passe"
                                value={password}
                                onChange={(e) => setPassword(e.target.value)}
                            />
                        </div>
                        {response?.message && <small className="danger">{response.message}</small>}
                        {loading ? (
                            <Spin data-testid="loading"/>
                        ) : (
                            <button
                                type="submit"
                                className="btn btn-custom btn-block"
                                disabled={loading}
                            >
                                Connexion
                            </button>
                        )}
                    </Form>
                </div>
            </div>

            <Footer/>
        </>
    );
}

export default Login;