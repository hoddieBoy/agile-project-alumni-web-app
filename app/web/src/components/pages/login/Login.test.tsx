import React from "react";
import Login from "components/pages/login/Login";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import LoginAction from "components/pages/login/Login.action";
import LoginLoader from "components/pages/login/Login.loader";
import axiosConfig from "config/axiosConfig";

// Mock axios configuration
jest.mock('config/axiosConfig');
const axiosConfigMock = axiosConfig as jest.Mocked<typeof axiosConfig>;

describe("The Login page functionality", () => {
    const router = createMemoryRouter(
        [
            {
                path: '/login',
                element: <Login />,
                loader: LoginLoader,
                action: LoginAction,
                children: []
            }
        ],
        {
            initialEntries: ['/login']
        }
    );

    describe("Render and basic elements", () => {
        beforeEach(() => {
            act(() => {
                render(<RouterProvider router={router} />);
            });
        });

        test("The Login page renders without crashing", () => {
            const tree = render(<RouterProvider router={router} />);
            expect(tree).toMatchSnapshot();
        });

        test("The Login page contains a header with the correct text", () => {
            expect(screen.getByText('Welcome to Alumni FIL')).toBeInTheDocument();
            expect(screen.getByText('Login to access administrative features')).toBeInTheDocument();
        });

        test("The Login page contains a form with the correct fields", () => {
            expect(screen.getByLabelText(/username/i)).toBeInTheDocument();
            expect(screen.getByLabelText(/password/i)).toBeInTheDocument();
        });

        test("The Login page contains a submit button", () => {
            expect(screen.getByRole('button', { name: /login/i })).toBeInTheDocument();
        });
    });

    describe("Given a non-authenticated user on the Login page", () => {
        beforeEach(() => {
            act(() => {
                render(<RouterProvider router={router} />);
            });
        });

        afterEach(() => {
            jest.restoreAllMocks();
        });

        test("When the user submits the form without entering any data, then an error message is displayed", async () => {
            const loginButton = screen.getByRole('button', { name: /login/i });

            userEvent.click(loginButton);

            expect(await screen.findByText('Username and password are required.')).toBeInTheDocument();
        });

        test("When the user submits the form with invalid data, then an error message is displayed", async () => {
            const usernameInput = screen.getByLabelText(/username/i);
            const passwordInput = screen.getByLabelText(/password/i);
            const loginButton = screen.getByRole('button', { name: /login/i });

            // Mock axiosConfig.post to return an error
            axiosConfigMock.post.mockRejectedValue({ response: { status: 401 } });

            userEvent.type(usernameInput, 'testuser');
            userEvent.type(passwordInput, 'password');
            userEvent.click(loginButton);

            await waitFor(() => {
                expect(screen.getByText('Invalid username or password.')).toBeInTheDocument();
            });
        });
    });
});