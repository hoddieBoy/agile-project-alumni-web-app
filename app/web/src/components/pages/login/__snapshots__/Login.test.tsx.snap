// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Login page functionality Render and basic elements The Login page renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="login-header"
      >
        <img
          alt="IMT Atlantique Logo"
          src="https://www.imt-atlantique.fr/sites/default/files/Images/Ecole/charte-graphique/IMT_Atlantique_logo_RVB_Negatif_Baseline_400x272.png"
        />
        <h3>
          Welcome to Alumni FIL
        </h3>
        <p>
          Login to access administrative features
        </p>
      </header>
      <div
        class="main-content"
      >
        <div
          class="login-container"
        >
          <form
            action="/login"
            id="login-form"
            method="post"
          >
            <div
              class="form-group"
            >
              <label
                for="username"
              >
                Username
              </label>
              <input
                class="form-control"
                id="username"
                name="username"
                placeholder="Enter your username"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="password"
              >
                Password
              </label>
              <input
                class="form-control"
                id="password"
                name="password"
                placeholder="Enter your password"
                type="password"
                value=""
              />
            </div>
            <button
              class="btn btn-custom btn-block"
              type="submit"
            >
              Login
            </button>
          </form>
        </div>
      </div>
      <footer
        class="login-footer"
      >
        <a
          href="#"
        >
          Contact Us
        </a>
         | 
        <a
          href="#"
        >
          Privacy Policy
        </a>
      </footer>
    </div>
    <div>
      <header
        class="login-header"
      >
        <img
          alt="IMT Atlantique Logo"
          src="https://www.imt-atlantique.fr/sites/default/files/Images/Ecole/charte-graphique/IMT_Atlantique_logo_RVB_Negatif_Baseline_400x272.png"
        />
        <h3>
          Welcome to Alumni FIL
        </h3>
        <p>
          Login to access administrative features
        </p>
      </header>
      <div
        class="main-content"
      >
        <div
          class="login-container"
        >
          <form
            action="/login"
            id="login-form"
            method="post"
          >
            <div
              class="form-group"
            >
              <label
                for="username"
              >
                Username
              </label>
              <input
                class="form-control"
                id="username"
                name="username"
                placeholder="Enter your username"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="password"
              >
                Password
              </label>
              <input
                class="form-control"
                id="password"
                name="password"
                placeholder="Enter your password"
                type="password"
                value=""
              />
            </div>
            <button
              class="btn btn-custom btn-block"
              type="submit"
            >
              Login
            </button>
          </form>
        </div>
      </div>
      <footer
        class="login-footer"
      >
        <a
          href="#"
        >
          Contact Us
        </a>
         | 
        <a
          href="#"
        >
          Privacy Policy
        </a>
      </footer>
    </div>
  </body>,
  "container": <div>
    <header
      class="login-header"
    >
      <img
        alt="IMT Atlantique Logo"
        src="https://www.imt-atlantique.fr/sites/default/files/Images/Ecole/charte-graphique/IMT_Atlantique_logo_RVB_Negatif_Baseline_400x272.png"
      />
      <h3>
        Welcome to Alumni FIL
      </h3>
      <p>
        Login to access administrative features
      </p>
    </header>
    <div
      class="main-content"
    >
      <div
        class="login-container"
      >
        <form
          action="/login"
          id="login-form"
          method="post"
        >
          <div
            class="form-group"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              class="form-control"
              id="username"
              name="username"
              placeholder="Enter your username"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              class="form-control"
              id="password"
              name="password"
              placeholder="Enter your password"
              type="password"
              value=""
            />
          </div>
          <button
            class="btn btn-custom btn-block"
            type="submit"
          >
            Login
          </button>
        </form>
      </div>
    </div>
    <footer
      class="login-footer"
    >
      <a
        href="#"
      >
        Contact Us
      </a>
       | 
      <a
        href="#"
      >
        Privacy Policy
      </a>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
